x-pack/plugins/lens/public/editor_frame_service/error_helper.ts

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+
+import { ExpressionRenderError } from 'src/plugins/expressions/public';
+
+interface RequestError extends Error {
+  body?: { attributes?: { error: { caused_by: { type: string; reason: string } } } };
+}
+
+const isRequestError = (e: Error | RequestError): e is RequestError => {
+  if ('body' in e) {
+    return e.body?.attributes?.error?.caused_by !== undefined;
+  }
+  return false;
+};
+
+export function getOriginalRequestErrorMessage(error?: ExpressionRenderError | null) {
+  return (
+    error &&
+    'original' in error &&
+    error.original &&
+    isRequestError(error.original) &&
+    i18n.translate('xpack.lens.editorFrame.expressionFailureMessage', {
+      defaultMessage: 'Request error: {causedByType}, {causedByReason}',
+      values: {
+        causedByType: error.original.body?.attributes?.error?.caused_by.type,
+        causedByReason: error.original.body?.attributes?.error?.caused_by.reason,
```

Comment:
The logic you're using here looks incomplete, so I tried to break it. There appear to be different error responses for different types of failures, and I think you've only tested permission failures. I don't feel comfortable with the sort of one-off error response handling that you're doing here unless we can validate it against some kind of reference error handling implementation. Maybe the ES Clients team can take a look at this?

For example, sometimes there is a `caused_by.caused_by` response from Elasticsearch when Elasticsearch fails to validate the request itself, such as when you use a terms agg on a `text` field. For example, try running this request in the dev tools:

```
GET kibana_sample_data_logs/_search
{
  "size": 0,
  "aggs": {
    "t": {
      "terms": {
        "field": "host"
      }
    }
  }
}
```

Another example of a response that you're missing a case for is when the script is failing. This can happen fairly often with an example like metricbeat:

```
GET metricbeat-*/_search
{
  "size": 0,
  "aggs": {
    "d": {
      "max": {
        "script": """doc['system.network.in.bytes'].value"""
      }
    }
  }
}
```

Reply:
I used the same logic as this piece of code from core: https://github.com/elastic/kibana/blob/24d29a31b8ee8d6eaa05cbd2c255350ef8b47148/src/core/public/notifications/toasts/error_toast.tsx#L71

I totally see your point, but I don’t think we have a “handleAllCommonEsErrors” helper (yet). What do you think about adding nice handling for the cases you mention above (great findings btw), and if none of these apply falling back to the default behavior of unknown errors (which is showing the message of the thrown exception and also what’s happening on master right now) - at least for 7.10. We should continue to investigate to find a holistic solution and ideally roll it out everywhere (Discover etc)